### DISCUSSION TITLE
Number of free parameters in an lmfit.Model

### DISCUSSION THREAD

**Discussion by eendebakpt** [lmfit/lmfit-py#825] [Q&A] [answered]
I would like to know the number of free parameters in a Model (used to check the input data for fitting has enough data points to perform the fitting). Is there functionality for this already?

If not, would adding a property like:
```
    @property
    def degrees_of_freedom(self) -> int:
        """ Return number of free parameters in the model """
        fixed_parameters = [ key for key, value in self.param_hints.items() if  value.get('vary', True) is False]
        return len(self.param_names) - len(fixed_parameters)
```
to the `lmfit.Model` be a good way to determine the number?

**Comment by reneeotten** [comment, marked as answer]
See the documentation, it’s already an attribute of the MinimizerResult —> nvarys 

https://lmfit.github.io/lmfit-py/fitting.html#lmfit.minimizer.MinimizerResult

Renee

**Reply by eendebakpt** [reply]
@reneeotten The `MinimizerResult` is only available after fitting right? I want to use the number of degrees of freedom to perform a check on the data before calling `Model.fit`

**Reply by newville** [reply]
Neither the number of data points nor the number of free parameters is (or can be) known by a `Model` until a fit is started.  

One might be able to count the number of variable parameters from a `Parameters` object, which would just be
```
nvarys = len([p for p in params.values() if p.vary])
```

That seems simple enough that I'm not sure that needs to be a property of `Parameters`, but I don't see a lot of harm in adding that.

It definitely does not belong in `Model`.

**Comment by newville** [comment]
@eendebakpt   See also https://lmfit.github.io/lmfit-py/model.html#lmfit.model.nfree: `nfree = ndata - nvarys` is the number of degrees of freedom for the fit.  Both `nvarys` and `ndata` are explicitly stated in every fit report.
